includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,386 @@
+              'type' => 'watchdog',
+              'severity' => WATCHDOG_ERROR,
+              );
+              $failureMessage['messages'][] = $messages;
+              watchdog_exception('islandora_video_streaming_panopto', $ex);
+              return $failureMessage;
+            }
+            $streaming['upload_id'] = $session['ID'];
+          } else {
+            $messages = array(
+              'message' => t('Issue with finishing upload for object: @pid, response: @response'),
+              'message_sub' => array(
+                '@pid' => $object->id,
+                '@response' => $response,
+              ),
+              'type' => 'watchdog',
+              'severity' => WATCHDOG_ERROR,
+            );
+            $failureMessage['messages'][] = $messages;
+          }
+        }
+      } catch (Exception $ex) {
+        $messages = array(
+          'message' => t('Panopto upload error: @exception'),
+          'message_sub' => array('@exception' => $ex->getMessage()),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+        watchdog_exception('islandora_video_streaming_panopto', $ex);
+      }
+    } else {
+      $messages = array(
+        'message' => t('Failed to retrieve session for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+    file_delete($archival_file['file'], true);
     if ($streaming) {
        // Write the data to a temporary file.
       $temp_name = drupal_tempnam('temporary://', 'file');
       if (file_put_contents($temp_name, json_encode($streaming)) === FALSE) {
-        $failureMessage['messages'][0]['message'] = t('Failed to create temporary file @filename for object @pid when uploading to Panopto.');
-        $failureMessage['messages'][0]['message_sub']['@filename'] = $temp_name;
+        $messages = array(
+            'message' => t("Failed to create temporary file @filename for object @pid when uploading to Panopto"),
+            'message_sub' => array(
+              '@pid' => $object->id, 
+              '@filename' => $temp_name
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
         return $failureMessage;
       }
-      return islandora_video_add_datastream($object, 'KALTURA_STREAM', $temp_name);
+      return islandora_video_add_datastream($object, 'PANOPTO_INFO', $temp_name);
     } else {
       return $failureMessage;
     }
   }
 }
 
+/**
+ * Requests token and returns token
+ * @param $oauthendpoint string The authentication URL for request
+ * @return string A token retrieved from server
```

Comment:
This return is actually string|null, as in:
`@return string|null A token retrieved from server, or null on failure`